packages/scene-system/sdk/SceneRuntime.ts

```diff
@@ -533,7 +538,30 @@ export abstract class SceneRuntime extends Script {
       }
 
       try {
-        await this.runCode(source as any as string, { dcl })
+        const { Permissions } = await this.loadAPIs(['Permissions'])
+
+        this.originalFetch = fetch
+        this.originalWebSocket = WebSocket
+
```

Comment:
I was looking at the logic inside the createWebSocket and createFetch functions and there is a flaw in the permissions. It runs asynchronously and therefore don't fail properly to create the fetch.

To prevent that async race condition it may be good to fetch the permissions in this function and to pass the booleans to the new functions:

```suggestion
const canUseWebsocket = await permission.hasPermission(PermissionItem.USE_WEBSOCKET)
```

Reply:
I haven't got the flaw yet and the logic in the creation methods are different. 
# Websockets 
I don't really like the websocket way, because the constructor runs always and the permission stops its functionality after the check, so in the meanwhile time everyone has allow to use web sockets. But, I think we can trust the async permission getting is faster than websocket connection.

# Fetch
I thought this implementation it was okay because the overwritten function it's a wrapped promise which checks before the permission and if it's okay continue the regular functionality. 

Pros:
We can ask or give to devs the possibility of asking users for permissions. For example when you enable your camera and the browser asking you to enable it. I thought something similar. 
Checking the permission before the scene loads, through away this possibility